### PR TITLE
[MRG] hnn-core GUI ERP partial tutorial

### PR DESCRIPTION
**Pull request by chenghuzi**
* Added first ERP tutorial (unfinished). 
* Exposed few methods used to manipulate the GUI

### CODE REVIEW

File: hnn_core/gui/gui.py [hunk 1]
```diff
@@ -518,6 +796,42 @@ def run_notebook_cells(self):
         """
         return js_string
 
+    # below are a series of methods that are used to manipulate the GUI
+    def _simulate_upload_file(self, file_url):
+        import urllib.request
```
Comment (jasmainak) [inline hnn_core/gui/gui.py:801]: import should be at the top of the file

File: hnn_core/gui/gui.py [hunk 2]
```diff
@@ -518,6 +796,42 @@ def run_notebook_cells(self):
...
+        tab_index = None
+        for idx in self.app_layout.left_sidebar._titles.keys():
+            if tab_title == self.app_layout.left_sidebar._titles[idx]:
+                tab_index = int(idx)
+                break
+        if tab_index is None:
+            raise ValueError("Incorrect tab title")
+        self.app_layout.left_sidebar.selected_index = tab_index
+
+    def _simulate_switch_plot_type(self, plot_idx, plot_type):
```
Comment (jasmainak) [inline hnn_core/gui/gui.py:830]: why are these private methods instead of public?

File: hnn_core/gui/gui.py [hunk 3]
```diff
@@ -28,6 +28,271 @@
...
+        grid[layout].children[plot_idx].children[0].value = plot_type
+        pass
+    elif layout == "U-D":
+        assert 0 <= plot_idx < 2
+        grid[layout].children[plot_idx].children[0].value = plot_type
+    else:
+        pass
+
+
+def initialize_viz_window(viz_grid, viz_window, simulation_data, plot_outputs,
```
Comment (jasmainak) [inline hnn_core/gui/gui.py:259]: This should be private, no?

File: hnn_core/gui/gui.py [hunk 4]
```diff
@@ -28,6 +28,271 @@
...
+            "new": default_plot_types[1]
+        })
+
+    grid = GridspecLayout(2, 1, height=layout.height)
+    grid[0, 0] = VBox([plot_dropdown_U, plot_outputs_U])
+    grid[1, 0] = VBox([plot_dropdown_D, plot_outputs_D])
+    return grid
+
+
+def change_plot_type(grid, layout, plot_idx, plot_type):
```
Comment (jasmainak) [inline hnn_core/gui/gui.py:246]: private

File: hnn_core/gui/gui.py [hunk 5]
```diff
@@ -518,6 +796,42 @@ def run_notebook_cells(self):
         """
         return js_string
 
+    # below are a series of methods that are used to manipulate the GUI
+    def _simulate_upload_file(self, file_url):
```
Comment (jasmainak) [inline hnn_core/gui/gui.py:800]: use this also in `test_gui.py` ?

**Comment by jasmainak** [comment]
@chenghuzi any updates here? When can we merge this tutorial?

**Comment by ntolley** [comment]
Isn't the goal to implement [this](https://jonescompneurolab.github.io/hnn-tutorials/erp/erp) tutorial verbatim? 

@jasmainak I think we need to walk through the tutorials ourselves before moving to merge this

**Comment by jasmainak** [comment]
I agree, I am waiting for @chenghuzi to tell us when it is ready from his end ... we need to really ramp up the pace. Still left to do:

- finish at least one tutorials
- changing the figure layout
- resizing GUI size
- miscellaneous improvements such as legacy_mode=False as default

**Comment by jasmainak** [comment]
@chenghuzi let's add the "analysis" tab in a separate PR please ...

**Comment by chenghuzi** [comment]
> @chenghuzi let's add the "analysis" tab in a separate PR please ...

But we need this for tutorial
image

**Comment by jasmainak** [comment]
The visualizations should not refresh any time a widget in the "analysis" tab changes, this is very distracting. We should discuss this in a separate PR

**Comment by jasmainak** [comment]
make a separate PR, get that merged and then rebase this PR :) Otherwise diff becomes hard to read

**Comment by jasmainak** [comment]
Also, if I want to start my tutorial, I won't have the latest GUI changes if they are not in master.

**Comment by chenghuzi** [comment]
> The visualizations should not refresh any time a widget in the "analysis" tab changes, this is very distracting. We should discuss this in a separate PR

I don't think so, because some parameters under analysis change the visualization results. For example,  smoothing window size changes the shape of dipole curve.

**Comment by chenghuzi** [comment]
> make a separate PR, get that merged and then rebase this PR :) Otherwise diff becomes hard to read

OK

**Comment by jasmainak** [comment]
I think a button called "update plot" should be used

**Comment by chenghuzi** [comment]
> I think a button called "update plot" should be used

That's not reactive

**Comment by jasmainak** [comment]
@chenghuzi @ntolley @rythorpe let's chat here: https://hnn-core.cloud.mattermost.com/main/channels/gsoc ... 

we need a chatroom to iterate fast

**Comment by chenghuzi** [comment]
> @chenghuzi @ntolley @rythorpe let's chat here: https://hnn-core.cloud.mattermost.com/main/channels/gsoc ...
> 
> we need a chatroom to iterate fast

It seems I can't sign up...

**Comment by jasmainak** [comment]
> That's not reactive

If you want to make it reactive, it should only update the line of the plot, not the entire visualization area. Right now you have an in-between solution. Doing a proper reactive solution will add complexity, which I'm not sure if it's worth it

**Comment by chenghuzi** [comment]
> > That's not reactive
> 
> If you want to make it reactive, it should only update the line of the plot, not the entire visualization area. Right now you have an in-between solution. Doing a proper reactive solution will add complexity, which I'm not sure if it's worth it

If we want to control the granularity of reactive update, that will be extremely complex. And now, it's just a single flash and everything will be updated, so I don't think it's very annoying compared with hitting a "update plot" button many times.

**Comment by jasmainak** [comment]
> It seems I can't sign up...

sign up with github? what do you see?

**Comment by chenghuzi** [comment]
> > It seems I can't sign up...
> 
> sign up with github? what do you see?
This is what I got:
image

I assume this is something like slack so we need username and password for each organization?

**Comment by jasmainak** [comment]
just sent an invite, does that help?

**Comment by chenghuzi** [comment]
I think it's ok now. I also changed several methods into private ones @jasmainak

**Comment by jasmainak** [comment]
I'll merge once the CIs come green!

**Comment by jasmainak** [comment]
The linkcheck is failing because of missing `nbsphinx` ...

File: hnn_core/gui/gui.py (hunk 5 above)
Comment (jasmainak) [reply hnn_core/gui/gui.py:800]: did you address this @chenghuzi ?

**Comment by chenghuzi** [comment]
> The linkcheck is failing because of missing `nbsphinx` ...

Just have it fixed. But we still got error as neuron.yale.edu is inaccessible. It seems this happened many times before and I tried to visit their website with safari and it does not work. Should we do something to this?

**Comment by jasmainak** [comment]
We should add that URL here:

https://github.com/jonescompneurolab/hnn-core/blob/1170bbc1e1c046128e4dfc8b3ae09ba1feb1d138/doc/conf.py#L205

File: hnn_core/gui/gui.py (hunk 5 above)
Comment (jasmainak) [reply hnn_core/gui/gui.py:800]: here is the part of the test I am imagining will become simpler: https://github.com/jonescompneurolab/hnn-core/blob/1170bbc1e1c046128e4dfc8b3ae09ba1feb1d138/hnn_core/tests/test_gui.py#L33-L50

**Comment by jasmainak** [comment]
@chenghuzi I left another comment inline with the code in case you missed it.

File: hnn_core/gui/gui.py (hunk 5 above)
Comment (chenghuzi) [reply hnn_core/gui/gui.py:800]: It's now fixed in our latest commit

**Comment by chenghuzi** [comment]
We've passed all tests, @jasmainak could you help to review and merge if it looks ok?

**Comment by jasmainak** [comment]
Perfect, thank you @chenghuzi 🥳 🥳 🥳 

What is the next step?

**Comment by chenghuzi** [comment]
> Perfect, thank you @chenghuzi 🥳 🥳 🥳
> 
> What is the next step?

To finish the analysis tab as well as the simulation loading part 😄

**Pull request merged**